### PR TITLE
Add authentication

### LINKED ISSUE #3: Application Owner should be able to sign in
https://www.youtube.com/watch?v=tMN7A5vrMv0

### PR DESCRIPTION
**Pull request by RasimKhusaenov**
## Summary

Resolves #3, #4.

## Test plan
<details><summary><h3>1. Sign Up</h3></summary>
<ol>
  <li>Go to http://localhost:3000/users/new</li>
  <li>You should see:
  </li>
  <li>After successful sign up, you should be redirected to the blog page:
  </li>
</ol>
</details>
<details><summary><h3>2. Sign In</h3></summary>
<ol>
  <li>Go to http://localhost:3000/sessions/new</li>
  <li>You should see:

</li>
  <li>After successful sign in, you should be redirected to the blog page: 

</li>
</ol>
</details>
<details><summary><h3>3. Sign Out</h3></summary>
<ol>
  <li>Go to any page and cllick on "Sign Out" button</li>
  <li>After successful sign out, you should be redirected to the blog page: 

</li>
</ol>
</details>

## Checklist:
- [x] I have performed a self-review of my own code
- [ ] I have added tests for new functionality
- [x] My pull request is shorter than 500 lines

### CODE REVIEW

File: db/schema.rb [hunk 1]
```diff
@@ -12,7 +10,20 @@
...
+ActiveRecord::Schema.define(version: 2022_03_27_131021) do
+
   # These are extensions that must be enabled in order to support this database
-  enable_extension 'plpgsql'
+  enable_extension "citext"
+  enable_extension "plpgsql"
+
+  create_table "users", force: :cascade do |t|
+    t.citext "email", null: false
+    t.string "password_digest"
```
Comment (Amir-Gaifullin) [inline db/schema.rb:21]:
```suggestion
    t.string "password_digest", null: false
```

File: db/migrate/20220326133631_create_users.rb [hunk 2]
```diff
@@ -0,0 +1,17 @@
+# frozen_string_literal: true
+
+class CreateUsers < ActiveRecord::Migration[6.0]
+  def change
+    enable_extension "citext"
+
+    create_table :users do |t|
+      t.citext :email, null: false
+      t.string :password_digest
```
Comment (Amir-Gaifullin) [inline db/migrate/20220326133631_create_users.rb:9]:
```suggestion
      t.string :password_digest, null: false
```

File: app/controllers/blog_controller.rb [hunk 3]
```diff
@@ -1,5 +1,8 @@
 # frozen_string_literal: true
 
 class BlogController < ApplicationController
+  before_action :authenticate_current_user!, except: %i[index]
```
Comment (DenisZackharov) [inline app/controllers/blog_controller.rb:4]: By rails style guide only options is more understandable then except: https://www.notion.so/flatstack/Style-Guide-425b94a54f7e481fa3f5dd10b58719ab#b525be45fe7f4ceb801362928f62e4ff

File: db/schema.rb (hunk 1 above)
Comment (RasimKhusaenov) [reply db/schema.rb:21]: Thx!

File: db/migrate/20220326133631_create_users.rb (hunk 2 above)
Comment (RasimKhusaenov) [reply db/migrate/20220326133631_create_users.rb:9]: Thx!

File: app/controllers/blog_controller.rb (hunk 3 above)
Comment (RasimKhusaenov) [reply app/controllers/blog_controller.rb:4]: Thx!

File: db/schema.rb (hunk 1 above)
Comment (Amir-Gaifullin) [reply db/schema.rb:21]: :sob:

**Review by Amir-Gaifullin — APPROVED** [review]
Good work!

File: Gemfile [hunk 4]
```diff
@@ -43,3 +43,5 @@ end
 gem "simple_form"
 gem "slim-rails"
 gem "tzinfo-data", platforms: %i[mingw mswin x64_mingw jruby]
+
+gem "pundit", "~> 2.2"
```
Comment (RamilyaNigmatullina) [inline Gemfile:47]: What do you think about adding `rubocop-gemfile` gem?

Comment (RasimKhusaenov) [reply Gemfile:47]: Good idea, but I don't think `rubocop-gemfile` is the right way to go. It is in beta, and last update was published 6 years ago.

Comment (RasimKhusaenov) [reply Gemfile:47]: In it's repo it says `USE Bundler/OrderedGems with rubocop v0.46.0+`.

File: app/models/user.rb [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+  has_secure_password
+
+  validates :email, presence: true, uniqueness: true, format: { with: URI::MailTo::EMAIL_REGEXP }
+  validates :first_name, presence: true
+
+  def to_s
+    full_name.presence || email
+  end
+
+  def full_name
```
Comment (RasimKhusaenov) [inline app/models/user.rb:13]: Use `UserDecorator` instead

File: app/views/layouts/application.html.erb [hunk 6]
```diff
@@ -13,6 +13,38 @@
...
 
   <body>
+    <nav class="navbar navbar-expand-lg navbar-light bg-light justify-content-between">
+      <a class="navbar-brand">My Own Blog</a>
+
+      <ul class="navbar-nav mr-auto">
+        <% if current_user.present? %>
+          <li class="nav-item">
+            <%= link_to(
+                  current_user.full_name.presence || current_user.email,
```
Comment (RasimKhusaenov) [inline app/views/layouts/application.html.erb:23]: As you have `to_s` method in `User` class with same code, you need to replace this.

**Pull request merged**